packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,67 +165,213 @@ function makeSwingStore(dirPath, forceReset = false) {
+                skipCount -= 1;
+              } else {
+                yield result;
+              }
+            } else {
+              break;
+            }
           }
+        } finally {
+          const statusEnd = streamStatus.get(streamName);
+          assert(
+            statusEnd === readStatus,
+            X`can't read stream ${q(streamName)}, it's been closed`,
+          );
+          closeStream(streamName);
+          assert(itemCount === 0, X`leftover item count ${q(itemCount)}`);
         }
-        fs.closeSync(fd);
-      }
-    } catch (e) {
-      if (e.code !== 'ENOENT') {
-        throw e;
       }
+      return reader();
     }
   }
 
   /**
-   * Append an item to a stream.
+   * Obtain a writer for a stream.
    *
-   * @param {string} streamName  The stream to append to
-   * @param {string} item  The item to append to it
+   * @param {string} streamName  The stream to be written
+   *
+   * @returns {StreamWriter} a writer for the named stream
    */
-  function appendItem(streamName, item) {
-    assert.typeof(
-      streamName,
-      'string',
-      X`non-string stream name ${streamName}`,
+  function openWriteStream(streamName) {
+    insistStreamName(streamName);
+    const status = streamStatus.get(streamName);
+    assert(
+      status === 'unused' || !status,
+      X`can't write stream ${q(streamName)} because it's already in use`,
     );
-    assert(streamName.match(/^[-\w]+$/));
-    let fd = streamFds.get(streamName);
-    if (!fd) {
-      fd = fs.openSync(`${dirPath}/${streamName}`, 'a');
-      streamFds.set(streamName, fd);
+    streamStatus.set(streamName, 'write');
+
+    // XXX fdTemp is a workaround for a flaw in TypeScript's type inference
+    // It should be fd, which it should be changed to when and if they fix tsc.
+    let fdTemp = streamFds.get(streamName);
+    if (!fdTemp) {
+      const filePath = `${dirPath}/${streamName}`;
+      const mode = fs.existsSync(filePath) ? 'r+' : 'w';
+      fdTemp = fs.openSync(filePath, mode);
+      streamFds.set(streamName, fdTemp);
     }
+    const fd = fdTemp;
     activeStreamFds.add(fd);
-    fs.writeSync(fd, item);
-    fs.writeSync(fd, '\n');
+
+    /**
+     * Write to a stream.
+     *
+     * @param {string} item  The item to write
+     * @param {Object|null} position  The position to write the item
+     *
+     * @returns {Object} the new position after writing
+     */
+    function write(item, position) {
+      assert.typeof(item, 'string');
+      assert(
+        streamFds.get(streamName) === fd,
+        X`can't write to closed stream ${q(streamName)}`,
+      );
+      if (!position) {
+        position = STREAM_START;
+      }
+      const buf = encoder.encode(`${item}\n`);
+      fs.writeSync(fd, buf, 0, buf.length, position.offset);
+      fs.fsyncSync(fd);
+      return {
+        offset: position.offset + buf.length,
+        itemCount: position.itemCount + 1,
+      };
+    }
+    return write;
```

Comment:
API usage question: when creating a vat (either from `initializeSwingSet` for static vats, or some flavor of `createVatDynamically`), we want to prepare the stream for writing, and save the `position` in the DB for later additions to the transcript, but we won't necessarily have anything to write at that moment. It feels like we would need one of:

* 1: find an excuse to write an initial record at the same time we first open the file, like `const writer = openWriteStream(vatID); pos = writer(initialSourceBundle)`, so later times we can always do the same thing: `pos = writer(transcriptEntry, pos)`
* 2: if we can't find such an excuse, write an empty record and just somehow skip the first one at replay time
* 3: or make the add-to-transcript function remember whether the stream has been opened or not and either call `pos = write(firstItem)` or `pos = write(laterItem, pos)` depending upon that flag.

None of those sound great. What about an API in which `create` takes a name and gives back an initial (zero) `pos` (but not a writer), and that's only ever called once per vat. Then a second `openForWriting` which takes a name and a mandatory `pos`, and returns the writer, and that's called every time the vat is paged in / brought online. ?

Reply:
oh.. or is `streamStore.STREAM_START` safe to use as the initial value of `pos`?

Reply:
I was planning on just initializing the position to `swingStore.START_POSITION`.